gpytorch/models/pyro/trace_predictive_log_likelihood.py

```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env python3
+
+import math
+from pyro.distributions.util import is_identically_zero
+from pyro.infer.elbo import ELBO
+from pyro.infer.enum import get_importance_trace
+from pyro.infer.util import MultiFrameTensor, get_plate_stacks, is_validation_enabled, torch_item
+from pyro.util import check_if_enumerated, warn_if_nan
+
+
+def _compute_log_r(model_trace, guide_trace):
+    log_r = MultiFrameTensor()
+    stacks = get_plate_stacks(model_trace)
+    for name, model_site in model_trace.nodes.items():
+        if model_site["type"] == "sample":
+            log_r_term = model_site["log_prob"]
+            if not model_site["is_observed"]:
+                log_r_term = log_r_term - guide_trace.nodes[name]["log_prob"]
+            log_r.add((stacks[name], log_r_term.detach()))
+    return log_r
+
+
+class Trace_PredictiveLogLikelihood(ELBO):
+    """
+    A trace implementation of predictive log likelihood-based inference.
+    This is similar to Pyro's Trace_ELBO, but where the ELBO is replaced by
+    the predictive log likelihood proposed by `Jankowiak et al., 2019`_.
+
+    .. note::
+
+        This loss function requires particles are vectorized.
+
+    .. _Jankowiak et al., 2019:
+        http://bit.ly/predictive_gp
+    """
+    def __init__(self, num_particles=1, vectorize_particles=True, max_plate_nesting=float('inf'), retain_graph=None):
+        if not vectorize_particles:
+            raise ValueError("Trace_PredictiveLogLikelihood requires that vectorize_particles=True")
+        super().__init__(
+            num_particles=num_particles,
+            vectorize_particles=True,
+            max_plate_nesting=max_plate_nesting,
+            retain_graph=retain_graph
+        )
+
+    def _get_trace(self, model, guide, *args, **kwargs):
+        """
+        Returns a single trace from the guide, and the model that is run against it.
+        """
+        model_trace, guide_trace = get_importance_trace("flat", self.max_plate_nesting, model, guide, *args, **kwargs)
+        if is_validation_enabled():
+            check_if_enumerated(guide_trace)
+        return model_trace, guide_trace
+
+    def loss(self, model, guide, *args, **kwargs):
+        """
+        :returns: returns an estimate of the predictive log likelihood
+        :rtype: float
+
+        Evaluates the predictive log likelihood with an estimator that uses num_particles many samples/particles.
+        """
+        loss = self.differentiable_loss(model, guide, *args, **kwargs)
+        return torch_item(loss)
+
+    def _differentiable_loss_particle(self, model_trace, guide_trace):
+        elbo_particle = 0
+        surrogate_elbo_particle = 0
+        log_r = None
+
+        # compute elbo and surrogate elbo
+        for name, site in model_trace.nodes.items():
```

Comment:
i'm not sure if this is what you want. what if you observe other data that has nothing to do with the GP?

Reply:
This is now removed.